`sprites` and `texts` functional pattern

just-do-halee opened the pull request:
I suggest giving this functionality `sprites` and `texts` variants.
I think it can perform better and be more efficient to write readable codes.

just-do-halee committed:
clippy leveled up

just-do-halee committed:
re-designing sprites and texts using methods pattern

just-do-halee committed:
its compatibility

just-do-halee committed:
cleaning code

CleanCut reviewed (commented):
I'm not in favor of most of these changes because I feel it makes it more difficult for me to understand.

I'm fine with the smaller improvements, like:
- the refactoring of the "if" expressions
- using functional style instead of `for` loops (where it is already possible).

I am not okay with the larger changes like:
- the `repository` macro
- the `Reposit` and `Entity` traits, and the resultant `BevyEntity` alias

You are, of course, welcome to maintain these own changes in your fork if you like, but I want to be really careful to keep rusty_engine in a state that is easy for me to understand.

If you would like to either make a new pull request, or alter this pull request, I'd be happy to accept the small changes that I mentioned!

just-do-halee committed:
reverts larger changes and preserves some useful methods on sprites and texts

just-do-halee commented:
Thank you for reviewing Nathan!
I deleted `repository` macro and also `Reposit` trait. 
But.. i preserved `Entity` trait, renamed to `EngineEntity` also removed `fn new(..)` method.
It has just `fn label(&self) -> &str;` trait method for achieving

> - the refactoring of the "if" expressions
> - using functional style instead of for loops (where it is already possible).

Adding small features and improving huge amount of user experience.

Now it has all of HashMap<String, T> methods and also

```rust
drain_all(&mut self) -> Map<_, fn((String, Entity)) -> Entity> {...}
drain_many(&mut self, mut predicate: impl FnMut(&&mut Entity) -> bool) -> Vec<Entity> {...}
drain_one(&mut self, predicate: impl FnMut(&&Entity) -> bool) -> Option<Entity> {...}
delete_many(&mut self, predicate: impl FnMut(&&mut Entity) -> bool) -> bool {...}
delete(&mut self, predicate: impl FnMut(&&Entity) -> bool) -> bool {...}
find_many(&self, predicate: impl FnMut(&&Entity) -> bool) -> Vec<&Entity> {...}
find(&self, predicate: impl FnMut(&&Entity) -> bool) -> Option<&Entity> {...}
find_mut_many(&mut self, predicate: impl FnMut(&&mut Entity) -> bool) -> Vec<&mut Entity> {...}
find_mut(&mut self, predicate: impl FnMut(&&mut Entity) -> bool) -> Option<&mut Entity> {...}
filter<O: FnMut(&&Entity) -> bool>(&self, predicate: O) -> Filter<Values<'_, String, Entity>, O> {...}
filter_mut<O: FnMut(&&mut Entity) -> bool>(&mut self, predicate: O) -> Filter<ValuesMut<'_, String, Entity>, O> {...}
for_each(&self, f: impl FnMut(&Entity)) {...}
for_each_mut(&mut self, f: impl FnMut(&mut Entity)) {...}
```

very useful methods too. no conflict.

CleanCut reviewed (changes requested):
Sorry for the delay, I've been swamped!

There is too much unnecessary complexity for me, still. The point of this library is to be very small and very simple for people learning Rust. I don't want to expose new concepts or methods with similar-but-different functionality to the others. The functional things you added in the other PR were okay since they were hidden in their own type behind an unusual method that folks won't likely notice or need to deal with.

Here is what I see that we need to change to get this merged:

- [ ] Remove the `repository` macro and the types generated by it
- [ ] Remove the `EngineEntity` trait
- [ ] Remove the `Text::new` method.
- [ ] Remove the `*_clod` methods
- [ ] Remove the `EngineRepo` struct

I realize that I'm essentially rejecting most of your changes. Sorry about that. 😞 Your changes do work, and I can see why you would find them useful, but I need to keep this library very focused on the Rust beginner learning experience.

Once again, it is perfectly fine if you would like to keep these changes in your own fork and use them yourself!
